# **Part IIB: Neural Networks**

In [2]:
import pandas as pd

In [4]:
data = pd.read_csv("MLF_GP1_CreditScore.csv")

# **Section IIIA: Investment Grade**

In [5]:
#Separate features and target variables for linear regression and logistic regression
X = data.iloc[:, : -2] #exclude target variables from the feature variables
y_inv = data.iloc[:, -2] #Select second-last column as target variable

In [6]:
from sklearn.model_selection import train_test_split
#Split train and test sets
X_train_inv, X_test_inv, y_train_inv, y_test_inv = train_test_split(X, y_inv, test_size=0.2, random_state=511)

In [7]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

In [9]:
#Create Neural Networks
model1 = Sequential([
    Dense(64, activation='relu', input_shape=(26,)),
    Dropout(0.2), #Use dropout feature to prevent overfitting
    Dense(32, activation='relu'),
    Dropout(0.2),
    Dense(1, activation='sigmoid')
])
model1.compile(optimizer="adam", loss='binary_crossentropy', metrics=['accuracy']) #The target is binary, so I will use crossentropy instead of mse
model1.fit(X_train_inv, y_train_inv, epochs=100)

Epoch 1/100
43/43 [==============================] - 6s 4ms/step - loss: 1.1694 - accuracy: 0.6846
Epoch 2/100
43/43 [==============================] - 0s 4ms/step - loss: 1.0437 - accuracy: 0.7074
Epoch 3/100
43/43 [==============================] - 0s 4ms/step - loss: 0.7624 - accuracy: 0.7368
Epoch 4/100
43/43 [==============================] - 0s 4ms/step - loss: 1.0187 - accuracy: 0.7449
Epoch 5/100
43/43 [==============================] - 0s 3ms/step - loss: 0.9357 - accuracy: 0.7419
Epoch 6/100
43/43 [==============================] - 0s 4ms/step - loss: 0.7502 - accuracy: 0.7500
Epoch 7/100
43/43 [==============================] - 0s 4ms/step - loss: 0.7585 - accuracy: 0.7507
Epoch 8/100
43/43 [==============================] - 0s 4ms/step - loss: 0.6850 - accuracy: 0.7522
Epoch 9/100
43/43 [==============================] - 0s 4ms/step - loss: 0.6432 - accuracy: 0.7471
Epoch 10/100
43/43 [==============================] - 0s 4ms/step - loss: 0.6098 - accuracy: 0.7603
Epoch 11/

In [10]:
print(model1.evaluate(X_test_inv, y_test_inv))

11/11 [==============================] - 0s 3ms/step - loss: 0.4385 - accuracy: 0.8235
[0.438455730676651, 0.8235294222831726]


The accuracy to predict whether firm is investment grade is 82.35%

# **Section IIIB: Credit Ratings**

In [33]:
#Select last column as target variable
y_rating = data.iloc[:, -1] 

In [34]:
from sklearn.preprocessing import OneHotEncoder
#Rating is categorical, so I use onehotencoding to convert it to numbers.
encoder = OneHotEncoder(sparse=False)
rating_encoded = encoder.fit_transform(y_rating.values.reshape(-1, 1))

/usr/local/lib/python3.9/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [35]:
#Convert encoded array to dataframe
y_rating = pd.DataFrame(rating_encoded, columns=encoder.get_feature_names_out(['Rating']))

In [37]:
#Split train and test sets
X_train_rating, X_test_rating, y_train_rating, y_test_rating = train_test_split(X, y_rating, test_size=0.2, random_state=511)

In [90]:
#Create Neural Networks
model2 = Sequential([
    Dense(128, activation='relu', input_shape=(26,)),
    Dropout(0.2),
    Dense(64, activation='relu'),
    Dropout(0.2),
    Dense(32, activation='relu'),
    Dropout(0.2),
    Dense(16, activation='softmax')
])
model2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy']) #The target is categorical, so I adjust loss argument
model2.fit(X_train_rating, y_train_rating, epochs=100, batch_size=10)

Epoch 1/100
136/136 [==============================] - 2s 3ms/step - loss: 3.7463 - accuracy: 0.1206
Epoch 2/100
136/136 [==============================] - 0s 4ms/step - loss: 2.9200 - accuracy: 0.1647
Epoch 3/100
136/136 [==============================] - 0s 4ms/step - loss: 2.6511 - accuracy: 0.1794
Epoch 4/100
136/136 [==============================] - 1s 4ms/step - loss: 2.5707 - accuracy: 0.1838
Epoch 5/100
136/136 [==============================] - 0s 4ms/step - loss: 2.5274 - accuracy: 0.1750
Epoch 6/100
136/136 [==============================] - 1s 4ms/step - loss: 2.4970 - accuracy: 0.1779
Epoch 7/100
136/136 [==============================] - 0s 3ms/step - loss: 2.4863 - accuracy: 0.2022
Epoch 8/100
136/136 [==============================] - 0s 4ms/step - loss: 2.4237 - accuracy: 0.1926
Epoch 9/100
136/136 [==============================] - 1s 5ms/step - loss: 2.4203 - accuracy: 0.2051
Epoch 10/100
136/136 [==============================] - 1s 5ms/step - loss: 2.3967 - accura

In [91]:
print(model2.evaluate(X_test_rating, y_test_rating))

11/11 [==============================] - 0s 4ms/step - loss: 2.4223 - accuracy: 0.3912
[2.4222590923309326, 0.3911764621734619]


The accuracy predicting credit rating of firm is 39.12%